vdc/k8s-config.yaml

```diff
@@ -271,3 +271,55 @@ spec:
   resources:
     requests:
       storage: 10Mi
+---
+apiVersion: v1
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  namespace: test
+  name: delete-test-pvc
+rules:
+- apiGroups: ["*"]
+  resources: ["persistentvolumeclaims"]
+  verbs: ["delete"]
+---
+apiVersion: v1
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: deploy-svc-delete-test-pvc
+  namespace: test
+subjects:
+- kind: ServiceAccount
+  name: deploy-svc
+  namespace: batch-pods
```

Comment:
Should this be in the same namespace as metadata: namespace (or omitted since defined under metadata)? At least in some of the documentation on RBAC, namespace is defined in metadata only.

Reply:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/

Reply:
I guess I'm a little unsure of what it means that subjects are in a different namespace than the Role they act on. Still researching. 

From the above link... ambiguous statements like this don't help: "A RoleBinding may reference a Role in the same namespace."

Reply:
I see, so the 'deploy-svc' service account is in the 'batch-pods' namespace; has nothing to do with the accessed resource's namespace.

Reply:
Following from that, if I understand correctly, the RoleBinding namespace is private the to the RoleBinding (which is a separate resource), and can be something different from the Role it refers to in roleRef. Is that right?

Reply:
Eh, documentation seems to suggest the RoleBinding's namespace is the scope that the subject can access. If so, can we omit the namespace properties altogether, when referencing a Role (as opposed to a ClusterRole, which has no namespace)?

Reply:
The answer is, apparently no. I deleted the role from the `test` namespace and added it to the `default` namespace and I could no longer delete PVCs in test.

```
(hail) dking@wmb16-359 # k auth can-i --as=system:serviceaccount:batch-pods:deploy-svc delete pvc -n test      
no
(hail) dking@wmb16-359 # k get rolebindings --all-namespaces                                 
NAMESPACE     NAME                                             AGE
batch-pods    batch-svc-batch-pods-admin-binding               117d
batch-pods    deploy-svc-delete-batch-pods-pvc                 45m
batch-pods    test-svc-batch-pods-admin-binding                117d
default       batch-svc-default-deploy-binding                 117d
default       batch-svc-test-test-binding                      117d
default       deploy-svc-default-deploy-binding                117d
default       deploy-svc-delete-test-pvc                       1m
default       gateway-read-letsencrypt-config                  117d
default       letsencrypt-update-letsencrypt-config            117d
default       notebook-create-services-and-pods                91d
kube-public   system:controller:bootstrap-signer               117d
kube-system   cloud-provider                                   117d
kube-system   fluentd-gcp-scaler-binding                       117d
kube-system   gce:cloud-provider                               35d
kube-system   heapster-binding                                 117d
kube-system   metrics-server-auth-reader                       117d
kube-system   system::leader-locking-kube-controller-manager   117d
kube-system   system::leader-locking-kube-scheduler            117d
kube-system   system:controller:bootstrap-signer               117d
kube-system   system:controller:cloud-provider                 117d
kube-system   system:controller:token-cleaner                  117d
test          test-svc-test-test-binding                       55m
```

If I create the RoleBinding in a namespace that is different from the Role it binds, I get an error:

```
(hail) dking@wmb16-359 # k get rolebindings -n foo                                                      
NAME                         AGE
deploy-svc-delete-test-pvc   1m
(hail) dking@wmb16-359 # k get roles -n foo       
No resources found.
(hail) dking@wmb16-359 # k auth can-i --as=system:serviceaccount:batch-pods:deploy-svc delete pvc -n foo    
no - role.rbac.authorization.k8s.io "delete-test-pvc" not found
```

I could probably make the Role a ClusterRole, but then I've increased its scope unnecessarily. Although there is repetition in namespace declaration, it seems better to limit everything to the namespace of intended use.

Reply:
Sounds good, thanks for the experiment. Definitely agree on not setting the ClusterRole. Only thing we haven't tried is not setting the namespace on a RoleBinding for a Role, but I think the explicit model is better, because it kind of seems that if we set the RoleBinding on a ClusterRole and don't specify a namespace, we will get cluster-wide access.